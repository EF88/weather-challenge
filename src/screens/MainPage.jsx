import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';

import { CityTypes } from '../store/cities/types';
import DayWeather from '../components/DayCard/DayCard';
import { WeatherSection, MenuCities, ButtonCities } from './MainPage.style';
import Spinner from '../components/SpinnerLoading/spinner.style';

const MainPage = () => {
  const dispatch = useDispatch();
  const { city: {consolidated_weather: days=[]}, cityRequest, cityFailure } = useSelector(
    state => state.cityState,
  );
  const menuButtons = [
    {
      name: 'Lisbon',
      code: '742676'
    },
    {
      name: 'London',
      code: '44418'
    },
    {
      name: 'New York',
      code: '2459115'
    },
    {
      name: 'Sao Paulo',
      code: '455827'
    }
  ];

  const handleOnClick = code => () => {
    dispatch({
      type: CityTypes.CITY_REQUEST,
      codeCity: code
    });
  }

  useEffect(() => {
    dispatch({
      type: CityTypes.CITY_REQUEST,
      codeCity: menuButtons[0].code
    });
  }, []);

  const renderNoWeather = () => {
    if (cityFailure) return <h1>Something went wrong to found the Weather</h1>;
    else if (cityRequest) return <Spinner />;
  };

  const renderDays = () => {
    return (
      <WeatherSection>
        {days && days.length > 0 && !cityRequest
         ? days.map( city => <DayWeather infoCity={city}/>)
         : renderNoWeather()
         }
      </WeatherSection>
    );
  };

  return (
    <>
      <MenuCities>
      { menuButtons.map(({name, code}) => <div><ButtonCities onClick={handleOnClick(code)}>{name}</ButtonCities></div>)}
      </MenuCities>
      {renderDays()}
    </>
  );
};

export default MainPage;
