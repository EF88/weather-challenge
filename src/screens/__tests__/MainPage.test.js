import React from 'react';
import { shallow } from 'enzyme';
import MainPage from '../MainPage';

jest.mock('react-redux', () => ({
useDispatch: ()=> { },
useSelector: () => ({
  city: { consolidated_weather: { days: [] } },
  cityRequest:false,
  cityFailure: false
})
}));

describe('<MainPage />', () => {
it('Expect to not log errors in console', () => {
const spy = jest.spyOn(global.console, 'error');
const wrapper = shallow(
<MainPage />,
);
expect(wrapper).not.toBeNull();
expect(spy).not.toHaveBeenCalled();
});
it('Should render and match the snapshot', () => {
const wrapper = shallow(
<MainPage />,
);
expect(wrapper).toMatchSnapshot();
});
});