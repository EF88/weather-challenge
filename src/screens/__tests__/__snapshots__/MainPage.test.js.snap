// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainPage /> Should render and match the snapshot 1`] = `ShallowWrapper {}`;
